New api

ericdevries opened the pull request:
Fixes DD-1341 and DD-1343

# Description of changes

Implemented the new API endpoints according to the spec. 

IMPORTANT: the database structure is now quite different, so if there is already data in the existing database we need to do some migrations @janvanmansum.

Also refactored the code a bit:
* removed unneeded interfaces
* deleted the duplicated models (DD-1343), all of them now end in Entity and have 1 Repository.
* moved all logic to Usecases class instead of inside services and resources
* renamed output packages for generated code to match dropwizard conventions
* removed duplicate html view
* added proper integration tests with inmemory database for some of the usecases

I suspect there is some merge which added some refactoring added to the the old code and this went unnoticed.

# How to test

# Open issues
* The OCFL object metadata reader uses jsonpath to read the JSONLD version of the oai-ore document, but it has not yet been tested with the version that the vault ingest flow generates. It would be wiser to implement it using Apache Jena (if that does not pose problems) or at least get the namespaces sorted out correctly.
* The service was designed to be dumb, but I believe that will cause problems with data being overwritten accidentally, and generating duplicate bag/version combinations because there is no locking. 
* Root response (the / endpoint) has a version, what is this version about?

(Add links)

*

# Notify

@DANS-KNAW/dataversedans

ericdevries committed:
Refactored to implement new API

ericdevries committed:
Improvements and simplification

ericdevries committed:
Fix tests, remove bloat, cleanup

ericdevries committed:
Cleanup

lindareijnhoudt commented on docs/api.yml:
```diff
@@ -107,7 +111,7 @@ paths:
           application/json:
             schema:
               oneOf:
```
the oneOf is a remainder of a previous version of this api. it can be removed

lindareijnhoudt commented on docs/api.yml:
```diff
@@ -118,11 +122,15 @@ paths:
+                $ref: '#/components/schemas/OcflObjectVersion'
         400:
           description: The OcflObject was syntactically incorrect, for instance not all required fields were provided.
         401:
```
the 409 conflict a few lines down is not needed if the service is not checking for conflicts

lindareijnhoudt reviewed (commented):
i only looked at the api.yml document

ericdevries committed:
Update input spec, optimize imports

jo-pol commented on src/main/java/nl/knaw/dans/catalog/core/UseCases.java:
```diff
@@ -0,0 +1,182 @@
+            .otherIdVersion(parameters.getOtherIdVersion())
+            .ocflObjectPath(parameters.getOcflObjectPath())
+            .filePidToLocalPath(parameters.getFilePidToLocalPath())
+            .exportTimestamp(parameters.getExportTimestamp())
```
I thought the convention for these method names was `with....`

jo-pol replied:
Oh, apparently generated with the Builder annotation.

ericdevries replied:
These are indeed openapi generated classes

ericdevries committed:
Rename parameter and remove old openapi spec

jo-pol commented on src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntity.java:
```diff
@@ -0,0 +1,93 @@
+@Setter
+@ToString
+@Builder
+@AllArgsConstructor
```
Does the DB use this? for manually written calls I would prefer a builder.

ericdevries replied:
Yes the builder uses it; however we can set it to package-private to not allow access to it from the outside

jo-pol commented on src/main/java/nl/knaw/dans/catalog/db/OcflObjectVersionEntityRepository.java:
```diff
@@ -0,0 +1,98 @@
+    @Override
+    public List<OcflObjectVersionEntity> findAllBySwordToken(String swordToken) {
+        return new ArrayList<>(
+            query("from OcflObjectVersionEntity where swordToken = :swordToken order by objectVersion desc limit 1")
```
findall yet retrieving just one? I suppose I don't understand.

ericdevries replied:
Good point, some leftovers I forgot to delete

jo-pol commented on src/main/java/nl/knaw/dans/catalog/db/TarEntity.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    public void setTarParts(List<TarPartEntity> tarParts) {
+        if (this.tarParts == null) {
```
Thought I saw some clever call for this somewhere.

jo-pol replied:
The memory resurfaces. It was about adding a(nother) value to a Map<String,List<...>>. That is another situation.

jo-pol commented on src/main/java/nl/knaw/dans/catalog/db/TarPartEntity.java:
```diff
@@ -0,0 +1,60 @@
+    private TarEntity tar;
+
+    @Override
+    public boolean equals(Object o) {
```
No annotation for this?

ericdevries replied:
Yes, but they recommend against doing this for JPA entities. See https://dzone.com/articles/lombok-and-jpa-what-may-go-wrong for more information.

jo-pol replied:
I would be tempted to change in ignorance. Comment everywhere might also be overkill. Don't know the balance.

ericdevries replied:
I have a plugin for IntelliJ by the creators of that article that warns me about that. It should also be possible to create some kind of linting rule that detects both @Entity and @Data on a class and rejects it.

jo-pol commented on src/main/resources/nl/knaw/dans/catalog/resource/view/error.ftl:
```diff
@@ -16,15 +16,15 @@
-                    <pre class="my-0"><code>${errorMessage.details}</code></pre>
-                </div>
-               </#if>
+                    <div class="bg-light my-5 py-4 px-4 border">
```
ftl is new to me

ericdevries replied:
It is a freemarker template, which allows for text to be interpolated with values. In this case it is a html document.

ericdevries replied:
See https://freemarker.apache.org/ for more info

ericdevries committed:
Add logging, reduce error verbosity

jo-pol reviewed (commented):
just some dumb question/remarks, skipped src/test

ericdevries committed:
Lombok on config

ericdevries committed:
Add unique constraint

ericdevries committed:
Added skeletonrecord, added metadata as object

ericdevries committed:
Made constructors package-private, fix broken queries

jo-pol reviewed (dismissed):
Don't want to block this work in progress.

ericdevries committed:
Update UUID parameters

ericdevries committed:
cleanup docs

ericdevries committed:
Minor updates

Renamed entities to just Tar etc
Renamed Repository implementation to DAO
Fixed issue with TAR -> OCFL object version ownership check
Removed archivalDate as required field

ericdevries committed:
Merge branch 'master' into new-api

ericdevries committed:
Add a test

ericdevries committed:
Add header
